polys: Improved speed of `get_dixon_matrix()` and revised tests

ctsiagkalis opened the pull request:
#### Brief description of what is fixed or changed

Updated `get_dixon_matrix()` so that it uses the max_degrees list
of the variables of the Dixon polynomial instead of those of the
polynomials passed to `__init__()`. This way, only the necessary
monomials needed to construct the Dixon matrix are computed.

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* polys
  * get_dixon_matrix() now computes only the necessary monomials for the Dixon matrix.
<!-- END RELEASE NOTES -->

ctsiagkalis committed:
polys: Improved speed of `get_dixon_matrix()` and revised tests

Updated `get_dixon_matrix()` so that it uses the max_degrees list
of the variables of the Dixon polynomial instead of those of the
polynomials passed to `__init__()`. This way, only the necessary
monomials needed to construct the Dixon matrix are computed.

jksuom commented on sympy/polys/multivariate_resultants.py:
```diff
@@ -126,13 +121,17 @@ def get_dixon_polynomial(self):
+        """
+        max_degrees = [
+            max(degree_list(Poly(poly, self.variables))[i] for poly
+                in polynomial.coeffs()) for i in range(self.n)]
```
It seems that, for each `i`, a new `Poly(poly, self.variables)` objects will be created for each `poly`, which is inefficient. I would first create all degree lists

    deg_lists = [degree_list(Poly(poly, self.variables))
                 for poly in polynomial.coeffs()]

and then compute the maxima for each `i`. That should be possible with something like this:

    max_degrees = [max(degs) for degs in zip(*deg_lists)]

(You can think of `deg_lists` as a matrix where the maximum of each column should be computed. `zip` will essentially transpose the matrix and turn the columns into rows `degs`.)

ctsiagkalis replied:
Oh, I see. Thanks for the review. I really liked your explanation for the code above. I'll fix it right away.

ctsiagkalis committed:
Improved the way `max_degrees` list is computed to be more efficient

This way, all degree lists are generated and stored in `deg_lists`
and then the maxima for each list is computed and stored in
`max_degrees` list.

jksuom commented:
This looks good but `get_upper_degree()` should not be removed yet as someone may be using it. It should be deprecated before removal (see [Deprecation policy](https://github.com/sympy/sympy/wiki/Deprecating-policy)). Alternatively, it could also be left as it is for now.

ctsiagkalis commented:
> 
> 
> This looks good but `get_upper_degree()` should not be removed yet as someone may be using it. It should be deprecated before removal (see [Deprecation policy](https://github.com/sympy/sympy/wiki/Deprecating-policy)). Alternatively, it could also be left as it is for now.

Keeping `get_upper_degree()` would require keeping the `max_degrees` instance attribute in `__init__()`, which I think is unnecessary and would complicate things, since it uses the same name as the list that `get_max_degrees()` returns. Also, since `get_upper_degree()` is a class method I'm not sure anyone would use it except for building the Dixon Matrix. Maybe it would be better to just keep only the new function and rename it to `get_upper_degree()`, since it replaces the functionality of the old function. Let me know what you think @jksuom .

jksuom commented:
Maybe we have to deprecate both `max_degrees` and `get_upper_degree()`. The attribute `max_degrees` would be replaced by a property that could be deprecated. Use the decorator [`@deprecated(...)`](https://github.com/sympy/sympy/blob/78d5a35e58c574fe76ee48f604147c09612ffc0f/sympy/core/decorators.py#L15) for both `max_degrees` and `get_upper_degree`.
See the docstring of [SympyDeprecationWarning](https://github.com/sympy/sympy/blob/78d5a35e58c574fe76ee48f604147c09612ffc0f/sympy/utilities/exceptions.py#L13) for the keyword parameters. It would suffice to set `feature`, `issue` and `deprecated_since_version="1.5"`. For `get_upper_degree`, also `use_instead="get_max_degrees"` could be set.

jksuom commented:
The attribute computed in `__init__()` could be hidden by renaming it `_max_degrees`. The deprecated `max_degree` property would then simply return `self._max_degrees`.

ctsiagkalis commented:
Thanks for clarifying that last part. Since this is the first time I am deprecating something, I am not totally sure how I should do it. This is what I came up with:


```
        # A list of the d_max of each variable.
        self._max_degrees = [max(degree_list(poly)[i] for poly in self.polynomials)
            for i in range(self.n)]

    @deprecated(feature="max_degrees",issue="17763",deprecated_since_version="1.5")
    @property
    def max_degrees(self):
        return self._max_degrees
```




```
@deprecated(feature="get_upper_degree",useinstead="get_max_degrees",issue=17763,deprecated_since_version="1.5")
    def get_upper_degree(self):
        list_of_products = [self.variables[i] ** self.max_degrees[i]
                            for i in range(self.n)]
        product = prod(list_of_products)
        product = Poly(product).monoms()

        return monomial_deg(*product)
```

And for the tests:
```
def test_get_upper_degree():
    """Tests upper degree function."""
    h = c * x ** 2 + y
    dixon = DixonResultant(polynomials=[h, q], variables=[x, y])

    with raises(SymPyDeprecationWarning):
        assert dixon.get_upper_degree() == 3
```

jksuom commented:
That looks fine to me. It is not necessary to add a regression test. (It will suffice to test `max_degrees` and `get_upper_degree()` locally to see that the code is OK.)

ctsiagkalis commented:
Ok, I'll remove the test. The use of the decorator seems to brake the code. For some reason `self` can't be passed as an argument to the functions. I will use the SymPyDeprecationWarning like this instead:

```
    @property
    def max_degrees(self):
        SymPyDeprecationWarning(feature="max_degrees",
                                issue=17763,
                                deprecated_since_version="1.5")
        return self._max_degrees
```
and
```
    def get_upper_degree(self):
        SymPyDeprecationWarning(feature="get_upper_degree",
                                useinstead="get_max_degrees",
                                issue=17763,
                                deprecated_since_version="1.5").warn()

        list_of_products = [self.variables[i] ** self.max_degrees[i]
                            for i in range(self.n)]
        product = prod(list_of_products)
        product = Poly(product).monoms()

        return monomial_deg(*product)
```

jksuom commented:
That does also look fine to me. I cannot think of why the decorator would break the code, but this should be OK if it will work better.

ctsiagkalis committed:
Deprecated `max_degrees` property and `get_upper_degree()`

Deprecated max_degrees property and get_upper_degree(). `get_max_degrees()` should be used instead.

jksuom commented:
> AssertionError: File contains trailing whitespace: /home/travis/build/sympy/sympy/sympy/polys/multivariate_resultants.py, line 139.

I have configured my editor (emacs) to show trailing whitespace in red. These errors can also be checked locally by `bin/test quality` if the editor does not show them.

ctsiagkalis committed:
Fixed trailing whitespace

ctsiagkalis commented:
Oh, my bad. I forgot to run the quality test. Gladly, it didn't use too much of the bot's time.

jksuom commented on sympy/polys/multivariate_resultants.py:
```diff
@@ -86,10 +85,16 @@ def __init__(self, polynomials, variables):
+    def max_degrees(self):
+        SymPyDeprecationWarning(feature="max_degrees",
+                        issue=17763,
+                        deprecated_since_version="1.5")
```
There should also be `.warn()`.

ctsiagkalis committed:
Fixed deprecation warning

jksuom commented on sympy/polys/multivariate_resultants.py:
```diff
@@ -127,37 +132,58 @@ def get_dixon_polynomial(self):
+                        issue=17763,
+                        deprecated_since_version="1.5").warn()
+
         list_of_products = [self.variables[i] ** self.max_degrees[i]
```
I now get two warnings from `get_upper_degree`, including one for `max_degrees`. This line could have `self._max_degrees[i]` to avoid the second warning.

ctsiagkalis replied:
I see. I didn't notice because I think the warnings are not displayed correctly to me.
Is that how the warning is supposed to look like?
```
get_upper_degree has been deprecated since SymPy 1.5. Use
get_max_degrees instead. See
https://github.com/sympy/sympy/issues/17763 for more info.

  deprecated_since_version="1.5").warn()
3
```

jksuom replied:
I think that the first three lines are the actual message. The last line is something that the interpreter will add. It is probably not useful but not harmful either.

ctsiagkalis committed:
Fixed warning in `get_upper_degree()`

Using `max_degrees` instead of `_max_degrees` caused a double warning.

jksuom commented:
Thanks, this is ready to be merged.
